lib/predictions/predictions.ex

```diff
@@ -142,23 +151,41 @@ defmodule Predictions.Predictions do
     :scheduled
   end
 
-  @spec sufficient_certainty?(map(), String.t()) :: boolean()
-  defp sufficient_certainty?(_stop_time_event, route_id)
-       when route_id in ["Mattapan", "Green-B", "Green-C", "Green-D", "Green-E"] do
-    true
+  defp get_destination_stop_id(trip_update) do
+    Enum.max_by(trip_update["stop_time_update"], fn update ->
+      if update["arrival"], do: update["arrival"]["time"], else: 0
+    end)
+    |> Map.get("stop_id")
   end
 
-  defp sufficient_certainty?(stop_time_event, _route_id) do
-    if Application.get_env(:realtime_signs, :filter_uncertain_predictions?) do
-      is_nil(stop_time_event["uncertainty"]) or stop_time_event["uncertainty"] <= 300
-    else
-      true
+  @spec get_prediction_type(String.t()) :: Prediction.prediction_type()
+  defp get_prediction_type(update_type) do
+    case update_type do
+      "mid_trip" -> :mid_trip
+      "at_terminal" -> :terminal
+      "reverse_trip" -> :reverse
+      _ -> nil
     end
   end
 
-  @spec has_departed?(Predictions.Prediction.t()) :: boolean()
+  defp is_valid_prediction?(stop_time_update) do
+    not (is_nil(stop_time_update["arrival"]) and is_nil(stop_time_update["departure"]) and
+           is_nil(stop_time_update["passthrough_time"]))
+  end
+
+  @spec is_excluded_prediction_type?(Prediction.t()) :: boolean()
+  defp is_excluded_prediction_type?(prediction)
+       when prediction.route_id in ["Mattapan", "Green-B", "Green-C", "Green-D", "Green-E"],
+       do: false
+
+  defp is_excluded_prediction_type?(prediction) do
+    if Application.get_env(:realtime_signs, :filter_uncertain_predictions?),
+      do: prediction.type in @excluded_prediction_types,
+      else: false
+  end
+
+  @spec has_departed?(Prediction.t()) :: boolean()
   defp has_departed?(prediction) do
-    prediction.seconds_until_departure && prediction.seconds_until_departure < 0 &&
-      not prediction.stopped_at_predicted_stop?
+    prediction.seconds_until_departure < 0 and not prediction.stopped_at_predicted_stop?
```

Comment:
nit: Technically this works, because `nil < 0 == false` according to Elixir's term ordering rules, but I think I prefer the explicit check as a reminder that `seconds_until_departure` might be `nil`